Public/Src/App/Bxl/AzureDevOpsListener.cs

```diff
@@ -0,0 +1,212 @@
+
+            switch (eventData.EventId)
+            {
+                case (int)EventId.PipProcessError:
+                    {
+                        var payload = eventData.Payload;
+
+                        m_buildViewModel.BuildSummary.PipErrors.Add(new BuildSummaryPipDiagnostic
+                        {
+                            SemiStablePipId = $"Pip{((long)eventData.Payload[0]):X16}",
+                            PipDescription = (string)eventData.Payload[1],
+                            SpecPath = (string)eventData.Payload[2],
+                            ToolName = (string)eventData.Payload[4],
+                            ExitCode = (int)eventData.Payload[7],
+                            Output = (string)eventData.Payload[5],
+                        }); 
+                    }
+                    break;
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void OnWarning(EventWrittenEventArgs eventData)
+        {
+            LogAzureDevOpsIssue(eventData, "warning");
+        }
+
+        /// <inheritdoc />
+        protected override void Output(EventLevel level, int id, string eventName, EventKeywords eventKeywords, string text, bool doNotTranslatePaths = false)
+        {
+        }
+
+        private void LogAzureDevOpsIssue(EventWrittenEventArgs eventData, string eventType)
+        {
+            var builder = new StringBuilder();
+            builder.Append("##vso[task.logIssue type=");
+            builder.Append(eventType);
+
+            var message = eventData.Message;
+            var args = eventData.Payload == null ? CollectionUtilities.EmptyArray<object>() : eventData.Payload.ToArray();
+            string body;
+
+            // see if this event provides provenance info
+            if (message.StartsWith(EventConstants.ProvenancePrefix, StringComparison.Ordinal))
+            {
+                Contract.Assume(args.Length >= 3, "Provenance prefix contains 3 formatting tokens.");
+
+                // file
+                builder.Append(";sourcepath=");
+                builder.Append(args[0]);
+
+                //line
+                builder.Append(";linenumber=");
+                builder.Append(args[1]);
+
+                //column
+                builder.Append(";columnnumber=");
+                builder.Append(args[2]);
+
+                //code
+                builder.Append(";code=DX");
+                builder.Append(eventData.EventId.ToString("D4"));
+            }
+
+            // report the entire message since Azure DevOps does not yet provide actionalbe information from the metadata.
+            body = string.Format(CultureInfo.CurrentCulture, message, args);
+
+            builder.Append(";]");
+
+            // substitute newlines in the message
+
+            var encodedBody = body.Replace("\r", "%0D").Replace("\n", "%0A");
+            builder.Append(encodedBody);
+
+            m_console.WriteOutputLine(MessageLevel.Info, builder.ToString());

```

Comment:
>MessageLevel.Info [](start = 38, length = 17)

all the messages are printed using `Info` level... this does not feel right

Reply:
At this point the info is only for color. I. Azuredevops mode there is no color.

Reply:
At this point the info is only for color. I. Azuredevops mode there is no color.

---
In reply to: [316891093](https://github.com/microsoft/BuildXL/pull/784#discussion_r316891093) [](ancestors = 316891093)